src/bin/pg_autoctl/fsm_transition.c

```diff
@@ -1189,3 +1230,307 @@ fsm_promote_standby(Keeper *keeper)
+									 WAIT_CASCADE_STATE,
+									 &(keeper->otherNodes)))
+		{
+			/* errors have already been logged */
+			return false;
+		}
+
+		if (keeper->otherNodes.count != 1)
+		{
+			log_error("Failed to find a node in state \"wait_cascade\" on "
+					  "the monitor, pgautofailover.get_other_nodes() "
+					  "returned %d node(s)",
+					  keeper->otherNodes.count);
+		}
+	}
+
+	/*
+	 * We expect a single WAIT_CASCADE node in the system at any point.
+	 */
+	upstreamNode = &(keeper->otherNodes.nodes[0]);
+
+	/*
+	 * Postgres 10 does not have pg_replication_slot_advance(), so we don't
+	 * support replication slots on standby nodes there.
+	 */
+	if (pgSetup->control.pg_control_version >= 1100)
+	{
+		strlcpy(slotName, config->replication_slot_name, MAXCONNINFO);
+	}
+
+	if (!standby_init_replication_source(postgres,
+										 upstreamNode,
+										 PG_AUTOCTL_REPLICA_USERNAME,
+										 config->replication_password,
+										 slotName,
+										 config->maximum_backup_rate,
+										 config->backupDirectory,
+										 upstreamNode->lsn,
+										 config->pgSetup.ssl,
+										 keeper->state.current_node_id))
+	{
+		/* can't happen at the moment */
+		return false;
+	}
+
+	if (!standby_fetch_missing_wal_and_promote(postgres))
+	{
+		log_error("Failed to fetch WAL bytes from standby node %d (%s:%d), "
+				  "see above for details",
+				  replicationSource.primaryNode.nodeId,
+				  replicationSource.primaryNode.host,
+				  replicationSource.primaryNode.port);
+		return false;
+	}
+
+	return true;
+}
+
+
+/*
+ * fsm_cleanup_and_resume_as_primary cleans-up the replication setting and
+ * start the local node as primary. It's called after a fast-forward operation.
+ */
+bool
+fsm_cleanup_and_resume_as_primary(Keeper *keeper)
+{
+	LocalPostgresServer *postgres = &(keeper->postgres);
+
+	if (!standby_cleanup_as_primary(postgres))
+	{
+		log_error("Failed to cleanup replication settings and restart Postgres "
+				  "to continue as a primary, see above for details");
+		return false;
+	}
+
+	if (!keeper_restart_postgres(keeper))
+	{
+		log_error("Failed to restart Postgres after changing its "
+				  "primary conninfo, see above for details");
+		return false;
+	}
+
+	return true;
+}
+
+
+/*
+ * When the selected failover candidate does not have the latest received WAL,
+ * it fetches them from another standby. This standby that has the latest WAL
+ * then goes in WAIT_CASCADE state to add the other node's HBA entries and
+ * such.
+ *
+ * The selected failover node is in state WAIT_FORWARD and we need to open an
+ * HBA entry for it. We already have a replication slot ready.
+ */
+bool
+fsm_prepare_cascade(Keeper *keeper)
+{
+	KeeperConfig *config = &(keeper->config);
+	Monitor *monitor = &(keeper->monitor);
+	LocalPostgresServer *postgres = &(keeper->postgres);
+	PostgresSetup *pgSetup = &(postgres->postgresSetup);
+
+	NodeAddress *candidateNode = NULL;
+
+	if (!config->monitorDisabled)
```

Comment:
We might have to finish that work someday, we're not in a position to remove it already, unfortunately. That said it's not fully supported, so what's important is to avoid breaking the case when we do have a monitor.